src/PersistentReducedStateStream.ts

```diff
@@ -0,0 +1,120 @@
+import {
+  BehaviorSubject,
+  Observable,
+  Subscription,
+  Subject,
+  Operator,
+  Subscriber,
+  TeardownLogic,
+} from 'rxjs';
+import { ActionStream } from './types/helpers';
+import { RegisteredReducer, combineReducers } from './reducer';
+import { defaultErrorSubject } from './internal/defaultErrorSubject';
+import { markName } from './internal/markers';
+import { tag } from 'rxjs-spy/operators';
+
+/**
+ * PersistentReducedStateStream is an Observable interface for state streams.
+ *
+ * The state stream has an inner subject which produces values to its subscribers.
+ *
+ * To begin reducing values, call `.startReducing(action$)`
+ * To stop reducing, call `.stopReducing()` (this closes the inner subject)
+ *
+ * The latest value of the state stream can be read synchronously by accessing
+ * `.state` (as long as the inner subject isn't closed)
+ */
+export class PersistentReducedStateStream<State> extends Observable<State> {
+  private subject: BehaviorSubject<State>;
+  private reducerSubscription?: Subscription;
+  public name: string;
+  private reducers: RegisteredReducer<State, any>[];
+  private errorSubject: Subject<any>;
+
+  /**
+   * Subscribes to the action stream and starts reducing this state stream.
+   * startReducing does not always immediately emit a value since the value
+   * might have already been emitted when the state stream was created.
+   *
+   * initialState can be passed to reset the state of the stream to some other value.
+   * This can be useful for hot-reloading
+   *
+   * startReducing will recreate the inner subject if it was previously closed
+   */
+  startReducing = (action$: ActionStream, initialState?: State) => {
```

Comment:
Should hot reloading actually reset the state? Is this desirable from a dev UX perspective?

Reply:
So far our goal has been for hot-reloading to keep the state (similar to how react-hot-loader works).

Reply:
Right, that's how I would prefer it too. I'm just wondering if there's any reason to pass initialState to this function.

Reply:
Aha, makes sense.

Having an argument for `initalState` is what allows us to avoid resetting the state when hot-reloading. When hot-reloading, the code that sets up the stream is re-ran, so if we didn't have a way to pass the "initial state", we wouldn't be able to insert the state that was built up before hot reloading (the stream would simply be reset to it's default state).

Perhaps the naming could be better, but the intention here is that `initialState` allows you to override what you use for the initial state of the stream. Normally the `initalState` is set to `defaultState`, however in the hot-reloading case `initialState` can be overriden to ensure that the stream state does not reset 😛  It's a bit of a mouthful.

Perhaps this code snippet demonstrates it better:

```typescript
// startStateStreams.ts
import action$ from 'action$';
import stateStream$ from 'somewhere';


// When hot-reloading module.hot.data.lastestStreamState will be set
// because the dispose handler is ran when this module was disposed
stateStream$.startReducing(action$, module.hot?.data?.latestStreamState);

if (module.hot) {
  module.hot.dispose(data => {
    data.latestStreamState = stateStream$.state;
    stateStream$.stopReducing();
  });
}
```

Reply:
Yeah, but don't we have `this.subject.value`? Maybe I am missing the point here :see_no_evil:

Reply:
`this.subject.value` can't be accessed when the stream has stopped reducing because it's closed (https://ncjamieson.com/closed-subjects/)

Reply:
Ok, got it now. Thanks.